src/charm.py

```diff
@@ -241,125 +223,135 @@ def _reconcile(self) -> None:  # pylint: disable=too-many-locals, too-many-state
-                )
-                self.unit.set_ports(80, 443)
-                if self.unit.is_leader():
-                    for backend in haproxy_route_requirers_information.backends:
-                        relation = self.model.get_relation(
-                            HAPROXY_ROUTE_RELATION, backend.relation_id
-                        )
-                        if not relation:
-                            logger.error("Relation does not exist, skipping.")
-                            break
-                        paths = (
-                            backend.application_data.paths if backend.path_acl_required else [""]
-                        )
-                        self.haproxy_route_provider.publish_proxied_endpoints(
-                            [
-                                f"https://{hostname}/{path}"
-                                for hostname in backend.hostname_acls
-                                for path in paths
-                            ],
-                            relation,
-                        )
-                    for (
-                        relation_id
-                    ) in haproxy_route_requirers_information.relation_ids_with_invalid_data:
-                        if relation := self.model.get_relation(
-                            HAPROXY_ROUTE_RELATION, relation_id
-                        ):
-                            self.haproxy_route_provider.publish_proxied_endpoints([], relation)
+                self._configure_haproxy_route(charm_state)
             case _:
                 if self.model.get_relation(TLS_CERT_RELATION):
                     # Reconcile certificates in case the certificates relation is present
                     tls_information = TLSInformation.from_charm(self, self.certificates)
                     self._tls.certificate_available(tls_information)
 
                 self.unit.set_ports(80)
-                self.haproxy_service.reconcile_default(config)
+                self.haproxy_service.reconcile_default(charm_state)
         self.unit.status = ops.ActiveStatus()
 
+    def _configure_ingress(self, charm_state: CharmState) -> None:
+        """Configure the ingress relation."""
+        tls_information = TLSInformation.from_charm(self, self.certificates)
+        self._tls.certificate_available(tls_information)
+
+        ingress_requirers_information = IngressRequirersInformation.from_provider(
+            self._ingress_provider
+        )
+        self.unit.set_ports(80, 443)
+        # In ingress mode it's expected for tls_information.hostnames
+        # to only contains the `external-hostname` charm config.
+        # Validation of tls_information will fail otherwise.
+        self.haproxy_service.reconcile_ingress(
+            charm_state, ingress_requirers_information, tls_information.hostnames[0]
+        )
+
+    def _configure_ingress_per_unit(self, charm_state: CharmState) -> None:
```

Comment:
You could merge this `_configure_ingress`. the only different I see is the attribute from which the `IngressRequirersInformation` is instantiated, and that can be parametrized